### PR TITLE
 eval derivative wrt modified #7437 b 

### PR DESCRIPTION
**Pull request by meghana-madhyastha**
@asmeurer , I added some minor changes to the original commits made by @clifford as you had suggested. I will close #10553. Please let me know if there is anything else to be added.

**TODOs**
- [ ] Tests for core functionality
- [ ] More direct tests (not just `isinstance`)
- [ ] Remove `new_name` parameter
- [ ] Use `xreplace` instead of `subs` in the default implementation
- [ ] Better method name
- [ ] Confusing return value names

### CODE REVIEW

**Comment by meghana1995** [comment]
@asmeurer  However, https://en.wikipedia.org/wiki/Matrix_calculus#Scalar-by-matrix suggested in issue #10509 hasn't been implemented in this PR (or in the previous commits). Should this be implemented as well ?

**Comment by asmeurer** [comment]
I will need some time to review this.

**Comment by asmeurer** [comment]
@meghana1995 your mess of commits makes it hard to see what you have changed beyond @cliffordwolf's branch. It looks like most of them are failed attempts at resolving merge conflicts, with bad commit messages. 

Can you please:

- Start with @cliffordwolf's branch (commit 3a762f5)
- Merge master into that branch, with a single merge commit (`git merge master`). If you want to write a commit message for this commit, add to the default one (don't use the `-m` to git).  This commit should *just* resolve the merge conflicts with master.
- Add your additions as new commits, with proper commit messages. See https://github.com/sympy/sympy/wiki/Development-workflow#writing-commit-messages for some tips on writing good commit messages. 

If you need help with the git stuff, feel free to ask on Gitter (it's easier to help with git in a real time chat).

**Comment by meghana1995** [comment]
@asmeurer . Sorry for the messy commits and the inconvenience caused as a result. I took your suggestions and added three commits: The first one which just merges with the current master, the second one to resolve merge conflicts and the third one with a couple of minor changes of mine. Thank you for the help.

**Comment by asmeurer** [comment]
Thanks, that's much better.

File: sympy/core/expr.py [hunk 1]
```diff
@@ -40,26 +40,65 @@ def _diff_wrt(self):
...
 
+    def _eval_derivative_wrt(self, expr, new_name):
+        """Transform derivative wrt a non-symbol variable to one wrt a symbol.
+
+        When taking the derivative of an expression with respect to (wrt) a
+        non-symbol variable with _diff_wrt=True, then this method is used to
+        create a new expression and a corresponding symbol variable.
+
+        Returns a tuple of new expr and symbol, or None if the transformation
+        could not be performed. The default implementation returns None.
```
Comment (asmeurer) [inline sympy/core/expr.py]: This needs to be updated.

**Comment by meghana1995** [comment]
@asmeurer , I changed the sentence in the doc.

**Comment by meghana1995** [comment]
@asmeurer , can this be reviewed ?

**Comment by leosartaj** [comment]
@meghana1995 Your branch has merge conflicts. Please update your branch to resolve them. :smile:

**Comment by meghana1995** [comment]
@leosartaj @asmeurer. I have updated my branch. Sorry for the delay and the merge conflicts.

**Comment by cliffordwolf** [comment]
Just out of curiosity: Is this going anywhere?
CI passes. Afaict there are no open objections to this patch. There are no merge conflicts.
Why is this not being merged?

File: sympy/core/expr.py [hunk 2]
```diff
@@ -40,26 +40,67 @@ def _diff_wrt(self):
...
+        ...         return (new_expr, new_self)
+        ...
+        >>> (x*MyClass()).diff(MyClass())
+        x
+        >>> (y*MyClass()).diff(MyClass())
+        Derivative(y*MyClass(), MyClass())
+        """
+        from sympy import Dummy
+        new_self = Dummy(new_name)
+        new_expr = expr.subs(self, new_self)
```
Comment (asmeurer) [inline sympy/core/expr.py]: A better default would probably be `xreplace`.

**Comment by asmeurer** [comment]
This could stand to have more tests.

File: sympy/core/expr.py [hunk 3]
```diff
@@ -40,26 +40,67 @@ def _diff_wrt(self):
...
+        ...         new_self = Dummy(new_name)
+        ...         new_expr = expr.subs(self, new_self)
+        ...         return (new_expr, new_self)
+        ...
         >>> (2*MyClass()).diff(MyClass())
         2
         """
         return False
 
+    def _eval_derivative_wrt(self, expr, new_name):
```
Comment (asmeurer) [inline sympy/core/expr.py]: From the comments on https://github.com/sympy/sympy/pull/7437, I don't think `new_name` is a needed argument here. The function can just use a `Dummy`, and pick its name to be whatever it wants (the name of a dummy doesn't matter, except for printing).

File: sympy/matrices/expressions/tests/test_matrix_exprs.py [hunk 4]
```diff
@@ -242,3 +243,8 @@ def test_Zero_power():
     assert z2**3 == MatPow(z2, 3).doit()
     assert z2**0 == Identity(3)
     raises(ValueError, lambda:MatPow(z2, -1).doit())
+
+
+def test_issue_7421():
+    assert isinstance(diff((D*w)[k,0], w[p,0]), Derivative)
+    assert isinstance(diff(w.T*D*w, w[0,0]), Derivative)
```
Comment (asmeurer) [inline sympy/matrices/expressions/tests/test_matrix_exprs.py]: Based on https://github.com/sympy/sympy/pull/7437#discussion_r12032369, this can't be tested directly because it uses a Dummy. It would be better to test the output directly somehow.

This is a bit worrisome, since it creates a whole class of expressions that can't be compared directly.

**Comment by asmeurer** [comment]
@cliffordwolf I just did a review of this, and I like it, except for these few comments. A lot of the comments are the same from the original https://github.com/sympy/sympy/pull/7437.

I don't know if @meghana1995 is still around to fix the issues. If not, I can make a new PR to implement them (unless someone else does first). 

Note that this has added the default implementation suggestion from https://github.com/sympy/sympy/pull/7437.

**Comment by asmeurer** [comment]
@Upabjojr this is related to a [message](https://groups.google.com/d/msg/sympy/UbadgQrrkZs/Lq3VmDqrAgAJ) you sent to the mailing list recently. Care to take a look?

**Comment by asmeurer** [comment]
In particular, can this be used to remove the hacks from https://github.com/sympy/sympy/pull/11144?

File: sympy/matrices/expressions/tests/test_matrix_exprs.py (hunk 4 above)
Comment (asmeurer) [inline sympy/matrices/expressions/tests/test_matrix_exprs.py]: I see, it's just the first one that uses Dummy, and it comes from the MatrixElement returning a Sum. So this is unrelated to this PR. To fix this (which would be a separate PR), I would suggest allowing MatrixElement to be unevaluated by default.

**Comment by asmeurer** [comment]
The method here could also use a better name. `_eval_derivative_wrt` doesn't actually evaluate the derivative.

**Comment by Upabjojr** [comment]
Had a quick glance. I don't see how to handle cases generating `KroneckerDelta` with `IndexedBase` with this PR, but maybe I'm missing something.

Better restart the tests anyway.

**Comment by asmeurer** [comment]
I don't see a point in restarting the tests when there are still several fixes necessary to make this branch mergable.

**Comment by asmeurer** [comment]
@Upabjojr how would you suggest solving this problem generically, i.e., some method that could go on objects that `diff` could call to sufficiently handle things like Indexed and MatrixElement derivatives?

File: sympy/core/expr.py [hunk 5]
```diff
@@ -40,26 +40,67 @@ def _diff_wrt(self):
...
+        """Transform derivative wrt a non-symbol variable to one wrt a symbol.
+
+        When taking the derivative of an expression with respect to (wrt) a
+        non-symbol variable with _diff_wrt=True, then this method is used to
+        create a new expression and a corresponding symbol variable.
+
+        Returns a tuple of new expr and symbol, or None if the transformation
+        could not be performed. The default implementation returns a tuple of
+        the 'new_expr' and 'new_self' where 'new_self' is a Dummy variable and
```
Comment (asmeurer) [inline sympy/core/expr.py]: These names are very confusing. In the function parameters, `self` is the object and `expr` is the thing we are taking a derivative with respect to, but here `new_expr` represents the modified `self` and `new_self` represents the replacement for `expr`.

File: sympy/matrices/expressions/matexpr.py [hunk 6]
```diff
@@ -339,6 +339,57 @@ class MatrixElement(Expr):
...
+        See Expr._eval_derivative_wrt for more information.
+        """
+        new_self = Dummy(new_name)
+
+        class FoundAmbigousExpression(Exception):
+            """Signal that we have found an Expression that may or may not be equal to self"""
+
+        def check_equal_index(a, b):
+            """Check if two MatrixElement indices are equal. Return None if result is ambigous."""
+            if a.equals(b):
```
Comment (asmeurer) [inline sympy/matrices/expressions/matexpr.py]: This is too smart. We should just use `==`.

File: sympy/core/function.py [hunk 7]
```diff
@@ -1150,23 +1155,14 @@ def __new__(cls, expr, *variables, **assumptions):
...
+                    obj = tmp_expr._eval_derivative(tmp_v) if tmp_expr is not None else None
                     if obj is not None:
-                        if not old_v.is_Symbol and obj.is_Derivative:
-                            # Derivative evaluated at a point that is not a
-                            # symbol
-                            obj = Subs(obj, v, old_v)
-                        else:
-                            obj = obj.xreplace({v: old_v})
-                    v = old_v
+                        obj = obj.subs(tmp_v, v)
```
Comment (asmeurer) [inline sympy/core/function.py]: Also here, why subs instead of xreplace?

**Comment by asmeurer** [comment]
In fact, looking more closely at the code, this doesn't even solve the problem for MatrixElement. It just returns None if the matrix is the same, causing the derivative to remain unevaluated. @cliffordwolf how would you use this to make `A[i].diff(A[j])` return `KroneckerDelta(i, j)`?

Maybe I'm missing something obvious, but this seems like a bad abstraction for fixing this problem. How should `A[i]` be "rewritten" so that its derivative wrt a dummy symbol is `KroneckerDelta(i, j)`? A more direct API (returning the derivative itself, instead of an expression to be differentiated) seems like it would be more useful.

**Comment by meghana1995** [comment]
@asmeurer @cliffordwolf Yes I can fix this. Give me a weeks time .

**Comment by Upabjojr** [comment]
> how would you use this to make A[i].diff(A[j]) return KroneckerDelta(i, j)?

If I remember correctly, I put `is_Symbol = True` in `Indexed`. There was still a problem with `A[i].has(A[j])` returning `False` (check was for optimization purposes, returning 0 to the derivative in that case). Don't remember right now how I solved it.

`isinstance(x, Symbol)` was replaced by `x.is_Symbol` somewhere.

I don't remember exactly now, but try maybe just putting `is_Symbol = True` in the matrix expression class.

**Comment by Upabjojr** [comment]
The trick to allow `A[i].diff(A[j])` is here: https://github.com/Upabjojr/sympy/blob/524043875ff55c7973f7b9e42bd0e3d37b0fb447/sympy/core/function.py#L1108

I don't think this problem is relevant for matrix symbols, as there are no indices bothering.

Concerning matrix symbols, I'd suggest to simply try to:

- put stuff like `is_Symbol = True` in the class definition.
- override `free_symbols` so as to return `{self}`

Maybe the solution to this problem is simpler than you think.

**Comment by asmeurer** [comment]
The problem is with MatrixElement, which is effectively the same as Indexed. 

Do you think a more general framework will be needed in the future to support things like matrix derivatives (https://github.com/sympy/sympy/issues/5858)?

Maybe all we really need is some replacement for `has`, like `_eval_derivative_depends_on` (except with a better name) to allow expressions to express that they depend on another expression for the purposes of differentiation. 

Also, is_Symbol checks in diff should be using `_diff_wrt`. That's an API that already exists.

**Comment by asmeurer** [comment]
Specifically, `free_symbols` doesn't cut it, because you can't make `A[i].free_symbols` include any `A[j]`.

**Comment by Upabjojr** [comment]
> Maybe all we really need is some replacement for has, like `_eval_derivative_depends_on` (except with a better name) to allow expressions to express that they depend on another expression for the purposes of differentiation. 

That's the point.

> Also, `is_Symbol` checks in diff should be using `_diff_wrt`. That's an API that already exists. 

So maybe we should simply correct the derivation algorithm and make sure that all expressions with `_diff_wrt = True` are locally treated as atoms (that is, do not enter their args while deriving).

> Do you think a more general framework will be needed in the future to support things like matrix derivatives (#5858)?

No, once you solve the aforementioned problems, I think defining `MatMul._eval_derivative` should be enough.

> Specifically, free_symbols doesn't cut it, because you can't make `A[i].free_symbols` include any `A[j]`.

That is the dirty trick I used to make `IndexedBase` derivatives work. I agree it should be undone once we solve the aforementioned problems.

**Comment by asmeurer** [comment]
> So maybe we should simply correct the derivation algorithm and make sure that all expressions with _diff_wrt = True are locally treated as atoms (that is, do not enter their args while deriving).

I recall hitting this issue before, I think with the vector module. 

> That is the dirty trick I used to make IndexedBase derivatives work. I agree it should be undone once we solve the aforementioned problems.

So to summarize:

- Add new API for expressions to tell if another, _diff_wrt expression is contained in it for the purposes of differentiation. The default implementation would use `has` (or should it be `free_symbols`?).
- Make the differentiation algorithm use _diff_wrt instead of is_Symbol checks. IIRC it gets away with symbol checks because it just replaces diff_wrt expressions with dummy symbols, but that sort of thing doesn't work in general (although I think it's what lead to the idea behind this PR). 
- Remove the hacks in diff around Indexed. Indexed (and similarly MatrixElement) should be able to get the behavior they want by implementing the API method from point 1 and setting themselves as `_diff_wrt = True`.

I'll open a new issue. I currently don't think this PR is a good idea (although anyone should feel free to argue otherwise).

**Comment by asmeurer** [comment]
https://github.com/sympy/sympy/issues/11568

Any objections to closing this PR?

**Comment by asmeurer** [comment]
I'm going to close this. PRs for https://github.com/sympy/sympy/issues/11568 are already starting to be merged.
